### ISSUE TITLE
Softmax similarity scores?

### ISSUE THREAD

**Issue by mholt** [asg017/sqlite-vec#115]
First, thank you for the excellent work on this module.

Is it possible to get softmax scoring?

I am new to this area, so apologies if I'm not using quite the right terms or am missing something obvious.

I have some Python code that does a vector similarity search manually: that is, without a database. I load in some data, discover the embeddings, and then get the embeddings for a search query that I then perform a distance calculation from with all the other embeddings.

The "similarity scores" are the dot product, then softmaxed; so that of all the results, all the scores are between 0-1, add up to 1, and scores of 0 mean "no match" while scores of 1 mean "identical."

I like this scring method because it is more intuitive for myself and other users. From what I can tell, the `distance` column returned by sqlite-vec is an arbitrary number.

I am wondering, could we somehow get softmax similarity scoring? I don't know if I has to be implemented into this module or if I can do it in my code -- if so, I'd love a pointer. :slightly_smiling_face:  Thanks!!

**Comment by asg017** [comment]
Hey @mholt!

By default the `distance` column is the L2 distance between the query vector (ie the `MATCH ?` vector) and the document vector. That distance can be configured with:

```sql
create virtual table vec_images using vec0(
  image_embedding float[1024] distance_metric=cosine
);
```

(valid values are `L1`/`L2`/`cosine`)

Does this match what you're thinking?

```sql

create virtual table vec_items using vec0(
    embedding float[4]
);

insert into vec_items(rowid, embedding) values 
    (1, '[0.1, 0.1, 0.1, 0.1]'),
    (2, '[0.2, 0.2, 0.2, 0.2]'),
    (3, '[0.3, 0.3, 0.3, 0.3]'),
    (4, '[0.4, 0.4, 0.4, 0.4]'),
    (5, '[0.5, 0.5, 0.5, 0.5]'),
    (6, '[0.6, 0.6, 0.6, 0.6]'),
    (7, '[0.7, 0.7, 0.7, 0.7]'),
    (8, '[0.8, 0.8, 0.8, 0.8]'),
    (9, '[0.9, 0.9, 0.9, 0.9]'),
    (10, '[1.0, 1.0, 1.0, 1.0]');

with knn as (
  select 
    rowid, 
    distance
  from vec_items
  where embedding match '[.5, .5, .5, .5]'
    and k = 6
)
select 
    *,
    exp(-distance) / sum(exp(-distance)) over () as softmax_score
from knn;

/*
┌───────┬───────────────────┬───────────────────┐
│ rowid │     distance      │   softmax_score   │
├───────┼───────────────────┼───────────────────┤
│ 5     │ 0.0               │ 0.220901075531133 │
│ 4     │ 0.199999988079071 │ 0.180858506081342 │
│ 6     │ 0.200000047683716 │ 0.180858495301335 │
│ 7     │ 0.399999976158142 │ 0.14807442264972  │
│ 3     │ 0.399999976158142 │ 0.14807442264972  │
│ 8     │ 0.600000023841858 │ 0.121233077786749 │
└───────┴───────────────────┴───────────────────┘
*/
```

Specifically that last line for `softmax_score` I think calculates what you want, but let me know if it doesn't match what you expect! I haven't heard of softmax scoring before, so it's useful I could try adding a custom SQL function for it

**Comment by mholt** [comment]
Ey, that works! Thank you!! I would have never come up with that myself.

I am also just realizing that I could do this after-the-fact, since it seems to be a post-processing of the raw distance values. Given that's the case, I could probably also compute this in my application :thinking: So maybe it doesn't need to be built into the module.

Is changing `distance_metric` supposed to show different values? I'd expect values between -1 and 1 for cosine similarity, but I still see values in like the 26 range.

Thank you for the great answer!

**Comment by mholt** [comment]
Oh actually... one more follow-up question for you:

Is there a way to return 0 results when nothing is relevant?

When I have 5 images in my database (of, say, trees/forests), and I search for "a glass of water", I get back all 5 results -- granted, with distance values of ~29 (L2 distance)... but relevant queries like "trees" have distance scores of 28. It's hard to know where to cut-off the results. I know it varies per-model, but they seem so close when talking about numbers on the order of 29.

Do you have any tips/advice? (I could also ask the discord, I guess, this isn't your responsibility to help with, heh.)

**Comment by asg017** [comment]
Unfortunately not, this is a general drawback of vector search. A KNN search will _always_ return K results, even if they are entirely irrelevant. Unless you combine it with full-text search [for hybrid search](https://alexgarcia.xyz/blog/2024/sqlite-vec-hybrid-search/index.html), then you can kindof reign things in, but that wouldn't work well if you're doing image embeddings. 

You could try to use the distance values themselves and make some custom thresholds, but that can be really tricky, ex a distance of "5" for one query can be exactly what you want, but absolutely wrong for another. It'll also depend on the embeddings models + queries, so hard to control. 

I think in production, people use multiple different systems to control their search results (fine-tuning models, re-rankers, OCR + separate object detection, user signals, custom models, etc.). But I don't have any good examples of those to show

(and asking here is totally fine! Github issues are more public anyway)

**Comment by mholt** [comment]
Ah, okay, that all makes sense. Thank you!

> I think in production, people use multiple different systems to control their search results (fine-tuning models, re-rankers, OCR + separate object detection, user signals, custom models, etc.).

That gives me the idea of filtering each result through the classifier relative to the search term and then sort by that score descending. This gives me scores of zero (well, with floating point imprecision, but basically 0) scores for irrelevant results, which I can then stop displaying those.

It's not particularly fast, but... the results are way better and relevant now!

Oh, and your hybrid search post is open in a tab on both my desktop and laptop since I will be implementing that into [Timelinize](https://github.com/timelinize/timelinize). :slightly_smiling_face:
